### PR TITLE
Fix the parsing of where-clauses for tuple-structs

### LINKED ISSUE #17904: Can't use `where` clauses with tuple structs
### STR

``` rust
// OK
struct Bar<T> where T: Copy {
    inner: T,
}

// Is this the right syntax?
struct Foo<T>(T) where T: Copy;  //~ error: expected `;`, found `where`

fn main() {}
```

### Version

``` rust
rustc 0.12.0-dev (10fc25a39 2014-10-09 16:04:25 -0500)
```

(PR 17669 on top of b08ecf7)

### PR DESCRIPTION
**Pull request by jroesch**
Fixes #17904. All the cases that I believe we should support are detailed in the test case, let me know if there is there is any more desired behavior. cc @japaric.

r? @nikomatsakis or whoever is appropriate.

### CODE REVIEW

**Comment by jroesch** [comment]
We might want to make this parsing code less flexible and reject code like: `struct Foo<T> where T: Copy (T) where T: Sized` which will parse unless we track to see if we have already parsed a where clause. We can't do this with a single look ahead AFAIK.

**Comment by sfackler** [comment]
I would personally prefer to reject where clauses in the old location entirely.

**Comment by nikomatsakis** [comment]
I agree with @sfackler, we should not accept `struct Foo<T> where T : Copy (T)` -- and in general I'd like to see some compile-fail tests with "weird" constructs that should not parse.

**Comment by nikomatsakis** [comment]
Travis is unhappy

/home/travis/build/rust-lang/rust/src/libsyntax/parse/parser.rs:4943: line longer than 100 chars

**Comment by nikomatsakis** [comment]
r+ once tidy error is fixed

**Comment by jroesch** [comment]
A small test failure was lurking, rebased, fixed it and pushed again.

File: src/test/run-pass/issue-17904.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// Copyright 2014 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
```
Comment (alexcrichton) [inline src/test/run-pass/issue-17904.rs]: Currently this file doesn't pretty print correctly: https://github.com/alexcrichton/rust/commit/acef2920d46350cbd9dc4f495072e3cc61bf7545

(it's fine to ignore pretty printing for now)

**Pull request merged**